README.md
# SpaceInvaders

Game developed as a part of Functional Programming Course, MiNI PW 2022

## Movement
- Move left/right using 'a' and 'd' respectively.
- Fire shoots using 'Space'.
- Navigate through menu and lost game window using mouse.

## Rules
- Player can move left/right but cannot move outside bounds of the game.
- Player has one life.
- Player looses life when either enemy missile hits him or any Invader reaches line on which spaceship operates.
- Invaders produce missiles with some probability which increases when player levels up.
- After each level the game starts again until player looses.
- Player can shoot missiles using 'Space'. Each invader has one life.
- Player can pause a game using 'P'
- Player can quit a game using 'Q'

ChangeLog.md
# Changelog for SpaceInvaders

## Unreleased changes


test/Spec.hs
import SpaceInvadersTests ( spaceInvadersTests )

main :: IO ()
main = spaceInvadersTests


test/SpaceInvadersTests.hs
module SpaceInvadersTests where
    import Test.QuickCheck ( quickCheck, Arbitrary, elements, chooseInt, choose, Gen, oneof, forAll, listOf, listOf1, NonEmptyList (getNonEmpty) )
    import Test.QuickCheck.Arbitrary (Arbitrary(arbitrary))
    import SpaceInvaders (Game (Game), GameAction(..), AppState (..), Spaceship (Spaceship), Invader (Invader), Missile (Missile), InvaderDirection (LeftDir, RightDir), startLevel, actionHandler, removeItem, missileRadius, isEmpty, Rectangle (Rectangle), Circle (Circle), recCircIntersect, removeIf, moveInvadersMissiles, moveInvaders, moveMissiles, moveSpaceshipRight, moveSpaceshipLeft)
    import System.Random ( mkStdGen, StdGen )

    spaceInvadersTests :: IO()
    spaceInvadersTests = quickCheck actionHandlerTest
        >> quickCheck isEmptyReturnsTrueIfEmpty
        >> quickCheck isEmptyReturnsFalseWhenNotEmpty
        >> quickCheck recCircIntersectTest
        >> quickCheck removeItemTest
        >> quickCheck moveInvadersMissilesRemovesMissileWhenIsOutOfBoundaries
        >> quickCheck moveInvadersMissilesRemovesMissileWhenHitSpaceship
        >> quickCheck moveMissilesRemovesMissileWhenIsOutOfBoundaries
        >> quickCheck moveMissilesRemovesMissileWhenHitInvader
        >> quickCheck moveInvadersDeletesInvaderWhenMissileHits
        >> quickCheck spaceshipStopsMovingWhenHitRightWall
        >> quickCheck spaceshipStopsMovingWhenHitLeftWall

    actionHandlerTest :: GameAction -> Game -> Bool
    actionHandlerTest Restart game@(Game Lost _ _ _ _ _ _ _ _ _ _) = actionHandler Restart game == startLevel Menu 1 0 (0, -270) [] []
    actionHandlerTest Restart game = actionHandler Restart game == game
    actionHandlerTest ga game@(Game Lost _ _ _ _ _ _ _ _ _ _) = actionHandler ga game == game
    actionHandlerTest Start game@(Game Menu spaceship invaders movs mis emis g c s p l) = actionHandler Start game == Game Running spaceship invaders movs mis emis g c s p l
    actionHandlerTest Start game@(Game Running _ _ _ _ _ _ _ _ _ _) = actionHandler Start game == game
    actionHandlerTest ga game@(Game Menu _ _ _ _ _ _ _ _ _ _) = actionHandler ga game == game
    actionHandlerTest Quit game = actionHandler Quit game == startLevel Menu 1 0 (0, -270) [] []
    actionHandlerTest PushLeft game@(Game Running spaceship invaders movs mis emis g c s p l) = actionHandler PushLeft game == Game Running spaceship invaders (movs ++ [PushLeft]) mis emis g c s p l
    actionHandlerTest PushRight game@(Game Running spaceship invaders movs mis emis g c s p l) = actionHandler PushRight game == Game Running spaceship invaders (movs ++ [PushRight]) mis emis g c s p l
    actionHandlerTest UnpushLeft game@(Game Running spaceship invaders movs mis emis g c s p l) = actionHandler UnpushLeft game == Game Running spaceship invaders (removeItem PushLeft movs) mis emis g c s p l
    actionHandlerTest UnpushRight game@(Game Running spaceship invaders movs mis emis g c s p l) = actionHandler UnpushRight game == Game Running spaceship invaders (removeItem PushRight movs) mis emis g c s p l
    actionHandlerTest Shoot game@(Game Running (Spaceship (x,y)) invaders movs mis emis g c s p l) = actionHandler Shoot game == Game Running (Spaceship (x,y)) invaders movs (mis ++ [Missile (x,y + 40) 10 missileRadius]) emis g c s p l
    actionHandlerTest Pause game@(Game Running spaceship invaders movs mis emis g c s p l) = actionHandler Pause game == Game Running spaceship invaders movs mis emis g c s (not p) l

    spaceshipStopsMovingWhenHitLeftWall :: Float -> Bool
    spaceshipStopsMovingWhenHitLeftWall spX
        | (spX - 5) < -455 = Spaceship (-455,0) == moveSpaceshipLeft (Game Running (Spaceship (spX, 0)) [] [PushLeft] [] [] (mkStdGen 0) 1 0 False 1)
        | otherwise = Spaceship (spX - 5,0) == moveSpaceshipLeft (Game Running (Spaceship (spX,0)) [] [PushLeft] [] [] (mkStdGen 0) 1 0 False 1)

    spaceshipStopsMovingWhenHitRightWall :: Float -> Bool 
    spaceshipStopsMovingWhenHitRightWall spX
        | (spX + 5) > 455 = Game Running (Spaceship (455,0)) [] [PushRight] [] [] (mkStdGen 0) 1 0 False 1 == moveSpaceshipRight (Game Running (Spaceship (spX,0)) [] [PushRight] [] [] (mkStdGen 0) 1 0 False 1)
        | otherwise = Game Running (Spaceship (spX + 5,0)) [] [PushRight] [] [] (mkStdGen 0) 1 0 False 1 == moveSpaceshipRight (Game Running (Spaceship (spX,0)) [] [PushRight] [] [] (mkStdGen 0) 1 0 False 1)

    moveInvadersDeletesInvaderWhenMissileHits :: Float -> Float -> Float -> Float -> Bool
    moveInvadersDeletesInvaderWhenMissileHits invX invY misX misY
        | recCircIntersect (Rectangle (invX, invY) 68 82) (Circle (misX, misY) 5) = null (fst (moveInvaders [Invader (invX, invY) invX LeftDir] [Missile (misX, misY) 10 5]))
        | otherwise = fst (moveInvaders [Invader (invX, invY) invX LeftDir] [Missile (misX, misY) 10 5]) /= []

    moveMissilesRemovesMissileWhenHitInvader :: Float -> Float -> Float -> Float -> Bool
    moveMissilesRemovesMissileWhenHitInvader invX invY misX misY
        | misY < 320  && recCircIntersect (Rectangle (invX, invY) 68 82) (Circle (misX, misY) 5) = null (moveMissiles [Missile (misX, misY) 10 5] [Invader (invX, invY) invX LeftDir])
        | misY < 320 && not (recCircIntersect (Rectangle (invX, invY) 68 82) (Circle (misX, misY) 5)) = moveMissiles [Missile (misX, misY) 10 5] [Invader (invX, invY) invX LeftDir] /= []
        | otherwise = null (moveMissiles [Missile (misX, misY) 10 5] [Invader (invX, invY) invX LeftDir])

    moveMissilesRemovesMissileWhenIsOutOfBoundaries :: Float -> Bool
    moveMissilesRemovesMissileWhenIsOutOfBoundaries misY
        | misY > 320 = null (moveMissiles [Missile (0, misY) 10 5] [])
        | otherwise = moveMissiles [Missile (0, misY) 10 5] [] /= []

    moveInvadersMissilesRemovesMissileWhenHitSpaceship :: Float -> Float -> Float -> Float -> Bool
    moveInvadersMissilesRemovesMissileWhenHitSpaceship emisX emisY spX spY
        | emisY > -320  && recCircIntersect (Rectangle (spX, spY) 36 58) (Circle (emisX, emisY) 5) = null (moveInvadersMissiles (Game Running (Spaceship (spX, spY)) [] [] [] [Missile (emisX, emisY) 10 5] (mkStdGen 0) 0 0 False 0))
        | emisY > -320 && not (recCircIntersect (Rectangle (spX, spY) 36 58) (Circle (emisX, emisY) 5)) = moveInvadersMissiles (Game Running (Spaceship (spX, spY)) [] [] [] [Missile (emisX, emisY) 10 5] (mkStdGen 0) 0 0 False 0) /= []
        | otherwise = null (moveInvadersMissiles (Game Running (Spaceship (spX, spY)) [] [] [] [Missile (emisX, emisY) 10 5] (mkStdGen 0) 0 0 False 0))

    moveInvadersMissilesRemovesMissileWhenIsOutOfBoundaries :: Float -> Bool
    moveInvadersMissilesRemovesMissileWhenIsOutOfBoundaries emisY
        | emisY < -320 = null (moveInvadersMissiles (Game Running (Spaceship (0, 0)) [] [] [] [Missile (0, emisY) 10 5] (mkStdGen 0) 0 0 False 0))
        | otherwise = moveInvadersMissiles (Game Running (Spaceship (10000, 10000)) [] [] [] [Missile (0, emisY) 10 5] (mkStdGen 0) 0 0 False 0) /= []

    removeItemTest :: Int -> [Int] -> Bool
    removeItemTest el list = el `notElem` removeItem el list

    recCircIntersectTest :: Rectangle -> Circle -> Bool
    recCircIntersectTest re@(Rectangle (x, y) w h) c@(Circle o r)
        | abs (fst o - x) > (w/2 + r) || abs (snd o - y) > (h/2 + r) = not (recCircIntersect re c)
        | abs (fst o - x) <= w/2 || abs (snd o - y) <= h/2 = recCircIntersect re c
        | otherwise = (cornerDistance <= r ** 2) == recCircIntersect re c
        where
            cornerDistance = (abs (fst o - x) - w/2) ** 2 + (abs (snd o - y) - h/2) ** 2

    isEmptyReturnsTrueIfEmpty :: Bool
    isEmptyReturnsTrueIfEmpty = isEmpty []

    isEmptyReturnsFalseWhenNotEmpty :: NonEmptyList Int -> Bool
    isEmptyReturnsFalseWhenNotEmpty list = not (isEmpty $ getNonEmpty list)

----------ARBITRARY----------
    instance Arbitrary GameAction where
        arbitrary = elements [Quit, PushLeft, PushRight, UnpushLeft, UnpushRight, Shoot, Pause, Start, Restart]

    instance Arbitrary Game where
        arbitrary = do
            appstate <- elements [Menu, Running, Lost]
            spaceshipX <- chooseFloat (-455, 455)
            invaderStartX <- chooseFloat (-400, 400)
            invaderX <- chooseFloat (invaderStartX - 10, invaderStartX + 10)
            invaderY <- chooseFloat (-320, 320)
            invaderDirection <- elements [LeftDir, RightDir]
            isPushLeft <- chooseBool
            isPushRight <- chooseBool
            misX <- chooseFloat (-455, 455)
            misY <- chooseFloat (-230, 330)
            invMisX <- chooseFloat (-320, 320)
            invMisY <- chooseFloat (-455, 455)
            randomGen <- chooseInt(1, 10000)
            invAttackChance <- chooseFloat(0, 1000)
            score <- chooseInt(0, 1000)
            pause <- chooseBool
            level <- chooseInt(0, 1000)
            return (Game appstate (Spaceship(spaceshipX, -270)) [Invader (invaderX, invaderY) invaderStartX invaderDirection] (addIf (addIf [] PushLeft isPushLeft) PushRight isPushRight) [Missile (misX, misY) 10 5] [Missile (invMisX, invMisY) 10 5] (mkStdGen randomGen) invAttackChance score pause level)

    instance Arbitrary Rectangle where
        arbitrary = do
            posX <- chooseFloat (-500, 500)
            posY <- chooseFloat (-500, 500)
            width <- chooseFloat (0, 500)
            height <- chooseFloat (0, 500)
            return $ Rectangle (posX, posY) width height

    instance Arbitrary Circle where
        arbitrary = do
            posX <- chooseFloat (-500, -500)
            posY <- chooseFloat (-500, 500)
            radius <- chooseFloat (0, 500)
            return $ Circle (posX, posY) radius

----------HELPERS----------
    chooseBool :: Gen Bool
    chooseBool = oneof [return True, return False]

    chooseFloat :: (Float, Float) -> Gen Float
    chooseFloat = choose

    addIf :: [a] -> a -> Bool -> [a]
    addIf list elt b = if b then list++[elt] else list


app/Main.hs
module Main where

import SpaceInvaders ( startLevel, update, AppState(Menu) )
import GlossRenderer
    ( update,
      glossEventToGameAction,
      keyHandler,
      renderGame,
      background,
      fps,
      window,
      ObjectBitmaps(..) )
import Graphics.Gloss (loadBMP, play, Display (InWindow), display, makeColor, Picture(Text))
import Graphics.Gloss.Data.Color

loadBitmaps :: IO ObjectBitmaps
loadBitmaps = do
  bkgImg <- loadBMP "./objects/background.bmp"
  shipImg <- loadBMP "./objects/spaceship.bmp"
  mstImg <- loadBMP "./objects/invader.bmp"
  return $ ObjectBitmaps bkgImg shipImg mstImg

main :: IO ()
main = 
  do
  imgs <- loadBitmaps
  play
    window
    background
    fps
    (startLevel Menu 1 0 (0, -270) [] [])
    (renderGame imgs)
    (keyHandler . glossEventToGameAction)
    update


src/GlossRenderer.hs
module GlossRenderer
  ( ObjectBitmaps(..)
  , renderGame
  , glossEventToGameAction
  , renderScores
  , keyHandler
  , update
  , module Window
  ) where

  import SpaceInvaders
    ( Game(spaceship, invaders, missiles, invadersMissiles, score, appstate),
      GameAction(..),
      Invaders,
      Invader(..),
      Spaceship(..),
      Missile(..),
      Missiles,
      InvadersMissiles,
      update,
      actionHandler,
      AppState(..) )

  import Window
  import Graphics.Gloss ( circleSolid, pictures, translate, Picture (Text), Color, color, black, red, scale, white, rectangleWire )
  import Graphics.Gloss.Interface.IO.Game
    ( white,
      scale,
      red,
      black,
      color,
      translate,
      pictures,
      circleSolid,
      SpecialKey(KeySpace),
      KeyState(Down, Up),
      Key(Char, MouseButton, SpecialKey),
      Event(EventKey),
      Color,
      Picture(Text),
      rectangleWire,
      MouseButton(LeftButton) )  

  data ObjectBitmaps = ObjectBitmaps
    { backgroundBmp :: Picture
    , spaceshipBmp :: Picture
    , invaderBmp :: Picture
    }

  renderGame :: ObjectBitmaps -> Game -> Picture
  renderGame objBmp game
    | appstate game == Menu =
    pictures
    [ renderStartRec
    , renderStartText
    , renderExitText
    ]
    | appstate game == Lost =
    pictures
    [ renderLostMessage
    , renderEndScore
    , renderRestartRec
    , renderRestartText
    , renderExitText
    ]
    | otherwise =
    pictures
    [ renderedBackground
    , renderedSpaceship
    , renderedInvaders
    , renderedMissiles
    , renderedInvadersMissiles
    , renderedScore
    ]
    where
        renderedBackground = renderBackground (backgroundBmp objBmp)
        renderedSpaceship
          = renderSpaceship (spaceship game) (spaceshipBmp objBmp)
        renderedInvaders
          = renderInvaders (invaders game) (invaderBmp objBmp)
        renderedMissiles = renderMissiles black (missiles game)
        renderedInvadersMissiles
          = renderMissiles red (invadersMissiles game)
        renderedScore = renderScores (450, 290) (0.2, 0.2) (score game)
        renderStartRec
          = color white (translate 0 200 (rectangleWire 800 50))
        renderStartText
          = scale
              0.4 0.4 (translate (- 150) 450 (color white (Text "Start")))
        renderExitText
          = scale
              0.4 0.4
              (translate
                 (- 600) (- 550) (color white (Text "Press 'Esc' to exit")))
        renderLostMessage
          = scale
              0.4 0.4 (translate (-300) 450 (color white (Text "You Lost!")))
        renderEndScore = renderScores (moveRenderScore $ score game, 0) (0.4, 0.4) (score game)
        renderRestartRec
          = color white (translate 0 (-100) (rectangleWire 800 50))
        renderRestartText
          = scale
              0.4 0.4
              (translate
                 (- 400) (-300) (color white (Text "Go to menu")))
        moveRenderScore score
          | score < 10 = 0
          | score >= 10 && score < 100 = -20
          | score >= 100 && score < 1000 = -40
          | score >= 1000 && score < 10000 = -60
          | otherwise = -80  

  renderBackground :: Picture -> Picture
  renderBackground bkgBmp = bkgBmp

  renderSpaceship :: Spaceship -> Picture -> Picture
  renderSpaceship (Spaceship (x, y)) = translate x y

  renderInvader :: Picture -> Invader -> Picture
  renderInvader bmp (Invader (x, y) _ _) = translate x y bmp

  renderInvaders :: Invaders -> Picture -> Picture
  renderInvaders invs bmp = pictures $ fmap renderInvaderWithImg invs
    where renderInvaderWithImg = renderInvader bmp

  renderScores :: (Float, Float) -> (Float, Float) -> Int -> Picture
  renderScores (x,y) (s1, s2) s = color white $ translate x y $ scale s1 s2 $ Text (show s)

  renderMissile :: Color -> Missile -> Picture
  renderMissile c (Missile (x,y) v r) = color c $ translate x y $ circleSolid r

  renderMissiles :: Color -> Missiles -> Picture
  renderMissiles c mis = pictures $ fmap (renderMissile c) mis

  glossEventToGameAction :: Event -> Maybe GameAction
  glossEventToGameAction (EventKey (MouseButton LeftButton) Up _ (x, y))
    | x <= 400 && x >= -400 && y <= 225 && y >= 175 =
    Just Start
    | x <= 400 && x >= -400 && y <= -75 && y >= -125 =
    Just Restart
    | otherwise = Nothing
  glossEventToGameAction (EventKey (Char 'q') Down _ _) = Just Quit
  glossEventToGameAction (EventKey (Char 'a') Down _ _) = Just PushLeft
  glossEventToGameAction (EventKey (Char 'a') Up _ _) = Just UnpushLeft
  glossEventToGameAction (EventKey (Char 'd') Down _ _) = Just PushRight
  glossEventToGameAction (EventKey (Char 'd') Up _ _) = Just UnpushRight
  glossEventToGameAction (EventKey (SpecialKey KeySpace) Down _ _) = Just Shoot
  glossEventToGameAction (EventKey (Char 'p') Down _ _) = Just Pause
  glossEventToGameAction _ = Nothing

  keyHandler :: Maybe GameAction -> Game -> Game
  keyHandler Nothing game = game
  keyHandler (Just k) game = actionHandler k game


src/SpaceInvaders.hs
module SpaceInvaders
    ( Spaceship (..)
    , Invader (..)
    , Invaders
    , GameAction (..)
    , Game (..)
    , Missile (..)
    , Missiles
    , InvadersMissiles
    , startLevel
    , update
    , actionHandler
    , missileRadius
    , AppState(..)
    , InvaderDirection(..)
    , removeItem
    , isEmpty
    , Rectangle(..)
    , Circle(..)
    , recCircIntersect
    , removeIf
    , moveInvadersMissiles
    , moveInvaders
    , moveMissiles
    , moveSpaceshipRight
    , moveSpaceshipLeft
    ) where

import System.Random ( uniformR, mkStdGen, StdGen )
import Graphics.Gloss.Interface.IO.Game ()
import Graphics.Gloss ()
import Data.Bifunctor ( Bifunctor(first) )
import GHC.Float ( int2Float )

----------BASE----------
data AppState = Menu | Running | Lost
  deriving (Eq, Show)

type World = (Float, Float)

----------HELPERS----------
data Rectangle = Rectangle World Float Float
  deriving (Eq, Show)

data Circle = Circle World Float
  deriving (Eq, Show)

----------SPACE INVADERS GAME STATE----------
data Missile = Missile
  { positionMis :: World
  , velocity :: Float
  , radius :: Float
  } deriving (Eq, Show)

missileRadius :: Float
missileRadius = 5

newtype Spaceship = Spaceship World
  deriving (Eq, Show)

data InvaderDirection = LeftDir | RightDir
  deriving (Eq, Show)

data Invader = Invader
  { positionInv :: World
  , startX :: Float
  , direction :: InvaderDirection
  }
  deriving (Eq, Show)

type Invaders = [Invader]

type Missiles = [Missile]

type InvadersMissiles = [Missile]

data GameAction = Quit | PushLeft | PushRight | UnpushLeft | UnpushRight | Shoot | Pause | Start | Restart
  deriving (Eq, Show)

data Game = Game
  { appstate :: AppState
  , spaceship :: Spaceship
  , invaders :: Invaders
  , movements :: [GameAction]
  , missiles :: Missiles
  , invadersMissiles :: InvadersMissiles
  , randomGenerator :: StdGen
  , invaderAttackChance :: Float
  , score :: Int
  , pause :: Bool
  , level :: Int
  }
  deriving (Eq, Show)

----------GAME ACTION HANDLERS----------
actionHandler :: GameAction -> Game -> Game
actionHandler Restart game@(Game Lost _ _ _ _ _ _ _ _ _ _) = startLevel Menu 1 0 (0, -270) [] []
actionHandler Restart game = game
actionHandler _ game@(Game Lost _ _ _ _ _ _ _ _ _ _) = game
actionHandler Start game@(Game Menu spaceship invaders movs mis emis g c s p l) = Game Running spaceship invaders movs mis emis g c s p l
actionHandler Start game@(Game Running _ _ _ _ _ _ _ _ _ _) = game
actionHandler _ game@(Game Menu _ _ _ _ _ _ _ _ _ _) = game
actionHandler Quit _ = startLevel Menu 1 0 (0, -270) [] []
actionHandler PushLeft (Game Running spaceship invaders movs mis emis g c s p l) = Game Running spaceship invaders (movs ++ [PushLeft]) mis emis g c s p l
actionHandler PushRight (Game Running spaceship invaders movs mis emis g c s p l) = Game Running spaceship invaders (movs ++ [PushRight]) mis emis g c s p l
actionHandler UnpushRight (Game Running spaceship invaders movs mis emis g c s p l) = Game Running spaceship invaders (removeItem PushRight movs) mis emis g c s p l
actionHandler UnpushLeft (Game Running spaceship invaders movs mis emis g c s p l) = Game Running spaceship invaders (removeItem PushLeft movs) mis emis g c s p l
actionHandler Shoot (Game Running (Spaceship (x,y)) invaders movs mis emis g c s p l) = Game Running (Spaceship (x,y)) invaders movs (mis ++ [Missile (x,y + 40) 10 missileRadius]) emis g c s p l
actionHandler Pause (Game Running spaceship invaders movs mis emis g c s p l) = Game Running spaceship invaders movs mis emis g c s (not p) l

----------GAME FUNCTIONS----------
startLevel :: AppState -> Int -> Int -> (Float, Float) -> [GameAction] -> InvadersMissiles -> Game
startLevel state l sc pos movs emis = Game
  { appstate = state
  , spaceship = Spaceship pos
  , invaders = [ Invader (-400, 280) (-400) LeftDir, Invader (-300, 280) (-300) LeftDir, Invader (-200, 280) (-200) LeftDir,
    Invader (-100, 280) (-100) LeftDir, Invader (0, 280) 0 LeftDir, Invader (100, 280) 100 LeftDir, Invader (200, 280) 200 LeftDir,
    Invader (300, 280) 300 LeftDir, Invader (400, 280) 400 LeftDir, Invader (-400, 200) (-400) LeftDir, Invader (-300, 200) (-300) LeftDir,
    Invader (-200, 200) (-200) LeftDir, Invader (-100, 200) (-100) LeftDir, Invader (0, 200) 0 LeftDir, Invader (100, 200) 100 LeftDir,
    Invader (200, 200) 200 LeftDir, Invader (300, 200) 300 LeftDir, Invader (400, 200) 400 LeftDir]
  , movements = movs
  , missiles = emis
  , invadersMissiles = []
  , randomGenerator = mkStdGen 2137
  , invaderAttackChance = int2Float (l - 1) * 0.01 + 0.1
  , score = sc
  , pause = False
  , level = l
  }

update :: Float -> Game -> Game
update _ game@Game {appstate = state, spaceship = Spaceship(t1,t2), invaders = invs, movements = movs, missiles = mis, invadersMissiles = emis, randomGenerator = g, invaderAttackChance = chance, score = sc, pause = p, level = l}
  | state == Menu || state == Lost || p = game
  | any (\x -> snd (positionInv x) <= -205) invs || any (\a -> recCircIntersect (Rectangle (t1,t2) 36 58) (Circle (positionMis a) 5)) emis =
  game {appstate = Lost}
  | isEmpty invs = startLevel Running (l + 1) sc (t1, t2) movs emis
  | otherwise = game {spaceship = moveSpaceship game, invaders = fst (moveInvaders invs mis), movements = movs, missiles = moveMissiles mis invs, invadersMissiles = moveInvadersMissiles game, randomGenerator = getNewGenerator invs g chance, score = sc + snd (moveInvaders invs mis)}

-- Moves spaceship left or right
moveSpaceship :: Game -> Spaceship
moveSpaceship g = moveSpaceshipLeft $ moveSpaceshipRight g

--Move spaceship left
moveSpaceshipLeft :: Game -> Spaceship
moveSpaceshipLeft game@(Game _ (Spaceship (x,y)) invs movs _ _ _ _ _ _ _) = if PushLeft `elem` movs then
                                                                      if (x - 5) < -455 then
                                                                        Spaceship (-455,y)
                                                                      else Spaceship(x-5,y)
                                                                    else Spaceship (x,y)

-- Move spaceship right, returns game for functions to fold
moveSpaceshipRight :: Game -> Game
moveSpaceshipRight game@(Game state (Spaceship (x,y)) invs movs mis emis g c s p l) = if PushRight `elem` movs then
                                                                            if (x + 5) > 455 then
                                                                              Game state (Spaceship (455,y)) invs movs mis emis g c s p l
                                                                            else Game state (Spaceship (x+5,y)) invs movs mis emis g c s p l
                                                                          else game

-- Move single invader
moveInvader :: Invader -> Invader
moveInvader (Invader (x,y) s dir)
                | x - 1 < s - 10 = Invader (x + 0.5, y - 4) s RightDir
                | x + 1 > s + 10 = Invader (x - 0.5, y - 4) s LeftDir
                | dir == LeftDir = Invader (x - 0.5, y) s LeftDir
                | dir == RightDir = Invader (x + 0.5, y) s RightDir
                | otherwise = error "Something went wrong"

-- Move invaders, delete if spaceship missile reached any, return new list of invaders and score
moveInvaders :: Invaders -> Missiles -> (Invaders, Int)
moveInvaders invs mis = first
  (moveInvader <$>)
  (removeIf
     (\ x
        -> any
             (\ y
                -> recCircIntersect
                     (Rectangle (positionInv x) 68 82) (Circle (positionMis y) 5))
             mis)
     invs)

-- Move single spaceship missile, return new missile
moveMissile :: Missile -> Missile
moveMissile (Missile (x,y) v r) = Missile (x, y + v) v r

-- Move spaceship missiles, delete if any are out of window or were used to kill invader
moveMissiles :: Missiles -> Invaders -> Missiles
moveMissiles mis invs = fst $ removeIf (\x -> snd (positionMis x) > 330) (moveMissile <$> fst (removeIf (\x -> any (\y -> recCircIntersect (Rectangle (positionInv y) 68 82) (Circle (positionMis x) 5)) invs) mis))

-- Move single invader missile
moveInvadersMissile :: Missile -> Missile
moveInvadersMissile (Missile (x,y) v r) = Missile (x, y - v) v r

-- Move invader missiles, delete if any are out of the window are destroyed spaceship
moveInvadersMissiles :: Game -> InvadersMissiles
moveInvadersMissiles game@(Game _ (Spaceship(x,y)) invs _ _ emis g chance _ _ _) = fst $ removeIf (\x -> snd (positionMis x) < -330) ((moveInvadersMissile <$> fst (removeIf (\a -> recCircIntersect (Rectangle (x, y) 36 58) (Circle (positionMis a) 5)) emis)) ++ getNewInvadersMissiles invs g chance)

-- Get new missiles after creating them
getNewInvadersMissiles :: Invaders -> StdGen -> Float -> InvadersMissiles
getNewInvadersMissiles invs g chance = fst (generateInvadersMissilesGenerator invs g chance)

-- Get new generator after creating invader missiles
getNewGenerator :: Invaders -> StdGen -> Float -> StdGen
getNewGenerator invs g chance = snd (generateInvadersMissilesGenerator invs g chance)

-- Given list of invaders, random generator and chance return list of generated missiles and new generator
generateInvadersMissilesGenerator :: Invaders -> StdGen -> Float -> (InvadersMissiles, StdGen)
generateInvadersMissilesGenerator [] g chance = ([], g)
generateInvadersMissilesGenerator (x:xs) g chance
  | null xs = ([Missile (positionInv x) 10 5 | fst (pseudoProbabilisticEvent chance g)], snd (pseudoProbabilisticEvent chance g))
  | otherwise = first
  (
     ([Missile (positionInv x) 10 5 |
         fst (pseudoProbabilisticEvent chance g)]) ++)
  (generateInvadersMissilesGenerator
     xs (snd (pseudoProbabilisticEvent chance g)) chance)

----------HELPER FUNCTIONS----------
-- Remove all given elements from list and return new list
removeItem :: Eq a => a -> [a] -> [a]
removeItem _ []     = []
removeItem x (y:ys) | x == y    = removeItem x ys
                    | otherwise = y : removeItem x ys

-- Remove all elements that fulfill given boolean expresion, return new list and number of deleted elements
removeIf :: (a -> Bool) -> [a] -> ([a], Int)
removeIf _ []     = ([], 0)
removeIf cond (y:ys)  | cond y    = (fst(removeIf cond ys), snd(removeIf cond ys) + 1)
                      | otherwise = first (y :) (removeIf cond ys)

-- Checks if Rectangle and Circle are intersecting (works only if rectangle is alligned with axis)
recCircIntersect :: Rectangle -> Circle -> Bool
recCircIntersect (Rectangle (x, y) w h) (Circle o r)
  | circleDistanceX > (w/2 + r) = False
  | circleDistanceY > (h/2 + r) = False
  | circleDistanceX <= w/2 = True
  | circleDistanceY <= h/2 = True
  | otherwise = cornerDistance <= r ** 2
  where
    circleDistanceX = abs (fst o - x)
    circleDistanceY = abs (snd o - y)
    cornerDistance = (circleDistanceX - w/2) ** 2 + (circleDistanceY - h/2) ** 2

-- With given probability of success and random generator returns Bool and new generator
pseudoProbabilisticEvent :: Float -> StdGen -> (Bool, StdGen)
pseudoProbabilisticEvent chance g = (fst (uniformR (1, 100000) g) <= (chance * 100), snd (uniformR (1 :: Float, 100000 :: Float) g))

-- Check if list is empty
isEmpty :: [a] -> Bool
isEmpty myList
  = case myList of
      [] -> True
      _ -> False
